### DISCUSSION TITLE
Adding ligatures and manipulating the components positions with a script

### DISCUSSION THREAD

**Discussion by khlsvr** [fontforge/fontforge#4904] [Q&A] [answered]
I have been looking at making a font with thousands of ligatures. I am able to add ligatures with the GUI manually, but for this process I need to automate it. I found a code example here https://stackoverflow.com/questions/10902593/fontforge-scripting-how-to-add-ligatures-for-a-glyph 

```
ligature_name = 'f_l'
ligature_tuple = ('f', 'l')
font.AddLookup('ligatures','gsub_ligature', (),[['rlig',[['arab',['dflt']]]]])
font.AddLookupSubtable('ligatures', 'ligatureshi')
glyph = font.createChar(-1, ligature_name)
glyph.addPosSub('ligatureshi', ligature_tuple)
```
Here's the process trying to add just this one ligature using scripting:

I downloaded just some test font that already has ligatures and it was this
https://www.dafont.com/antro-vectra.font
 Then I just used the method for adding ligatures described in this tutorial https://www.youtube.com/watch?v=UUUeogQAjv0 and adde one ligature for a_b. Generated a font file, at which point it didn't like the em size was 1300 and that it should be 1000. I continued to generate the font, generated as an otf font. It also found errors: self intersecting, missing points at extrema, too many points. It allowed me to still generate the font and I did. Then I installed it and tested it, and my newly created ligature worked. All good so far. Then I created the above mentioned script but modified the names in it to "liga" both. Then I opened fontforge and try edit the fonts lookup table and subtable names for ligatures to "liga" and generate the font again, there is something going on, because then after I open the newly generated font, the old default name `'liga' Standard Ligatures in Latin lookup 1 subtable` is back (and not the simple "liga" that I edited there. Anyway if the function font.addLookup() and font.addLookupSubtable() are supposed to add new tables then the names shouldn't match anyway right?
 
Anyway this is what the file may look after simplifying: 

``` 
#!/usr/bin/fontforge

ligature_name = 'f_l'
ligature_tuple = ('f', 'l')
font.AddLookup('liga','gsub_ligature', (),[['rlig',[['liga',['dflt']]]]])
font.AddLookupSubtable('liga', 'liga')
glyph = font.createChar(-1, ligature_name)
glyph.addPosSub('liga', ligature_tuple) 
```
Then I run it and get this error:

```
./make_liga.pe ~/.local/share/fonts/AntroVectra1.otf             
Copyright (c) 2000-2021. See AUTHORS for Contributors.
 License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
 with many parts BSD <http://fontforge.org/license.html>. Please read LICENSE.
 Version: 20201107
 Based on sources from 2021-12-02 19:17 UTC-ML-D-GDK3.
./make_liga.pe: 4 Expected rparen, got comma
```

Note I changed the `arab` to `liga`. I tried `latin` there too but same error happens. This is what the manual says about the function https://fontforge.org/docs/scripting/scripting-alpha.html#AddLookup
The `4` in error indicates line before that though I guess.

**Comment by ctrlcctrlv** [comment, marked as answer]
If you're going to be learning scripting in FontForge, it's non-intuitive to start with the native scripting language. Per [§D2](https://github.com/fontforge/fontforge/wiki/Community-guidelines#D2) (emphasis mine):

> FontForge includes two scripting capabilities: first, via a [legacy scripting language](https://fontforge.org/docs/scripting/scripting.html) created by George Williams, FontForge's original author, called internally `ff`. The other is Python 3.
> 
> **We recommend no new scripts in the legacy language be written. FontForge's Python API is always to be used for new scripts.**
> 
> It is allowed, as of at least [PR &numero;4154](https://github.com/fontforge/fontforge/pull/4154), to implement a feature in the Python API and not backport it to the legacy language.
> 
> There are no plans to remove the legacy language, and we of course still accept pull requests fixing issues in it and bringing it up to parity with Python. However, a request to add a feature to it that's *not* in Python *must* add it to Python as well to be accepted.
> 
> Everything possible in the GUI, within reason, should also be possible in Python, so it's also usually not acceptable to add a feature _only_ to the GUI.

**Reply by ctrlcctrlv** [reply]
By the way it looks like you're very confused and trying to _mix_ scripting languages. Half your code looks to me like legacy `ff`, half like Python. Move your file to `.py` and write it all in Python, doing away with `AddLookup`. Here is some code [from the Python docs](https://fontforge.org/docs/scripting/python/fontforge.html)…

```python3
# Add a `mark` lookup for an arbitrary glyph...
script = fontforge.scriptFromUnicode(glyph.unicode)
font.addLookup("l1", "gpos_mark2base", None, (("mark",((script,("dflt")),)),))
font.addLookupSubtable("l1", "l1-1")
font.addAnchorClass("l1-1", "top")
```

**Reply by khlsvr** [reply]
Thanks for the quick reply! I didn't pay much attention to the quality of the snippet I was using. It was at stackoverflow and had some upvotes. 

I'm a very mediocre programmer myself, and was wondering since adding ligatures is quite normal task that I could find an example on how to do it but if that's not the case I will try read and understand how to do it using the docs. I don't have a good track record on these things but I'll give it a shot tomorrow or so

**Reply by ctrlcctrlv** [reply]
That's not a safe assumption with FontForge. The SO user seemed to be assuming a programmer of some competence, they don't even load a font. I edited their answer a bit, but you can't just c&p from SO and expect anything to work—that's what sharing scripts/new Python plugin system is for, completed scripts.

![image](https://user-images.githubusercontent.com/838783/152246005-1dd10f0e-ff3d-4c10-8c77-f30b2e111f2f.png)

**Reply by khlsvr** [reply]
Great! I'm getting somewhere. With this I can save an sfd file which shows the created table and subtable, but there is no f_l in the subtable for some reason. The otf file however doesn't have the ligature and ligatureshi table+subtable. Maybe I need to prepare and do something more. I'll continue tomorrow thanks again.

```
#!/usr/bin/env python3
import fontforge

# load your font, etc…
font = fontforge.open("ant.otf")

ligature_name = 'f_l'
ligature_tuple = ('f', 'l')
font.addLookup('ligatures','gsub_ligature', (),[['liga',[['latn',['dflt']]]]])
font.addLookupSubtable('ligatures', 'ligatureshi')
glyph = font.createChar(-1, ligature_name)
glyph.addPosSub('ligatureshi', ligature_tuple)

font.save('ant.sfd')
font.generate('ant1.otf')
```

**Reply by ctrlcctrlv** [reply]
I think it's because you're not putting `'opentype'` flag in `font.generate`

**Reply by khlsvr** [reply]
Update + 1 question about positioning components. There are some problems, but they are not important for now (I think), but I'll show them too in the bottom.

The important part now is that I want to know how to move the ligature reference components so that they are on top of each other instead of being next to each other like they are by default. Example code is best, but even if I know what functions I should look at.

```
# load your font, etc…
font = fontforge.open("ant.otf")

ligature_name = 'f_y'
ligature_tuple = ('f', 'y')
font.addLookup('ligatures','gsub_ligature', (),[['liga',[['latn',['dflt']]]]])
font.addLookupSubtable('ligatures', 'ligatureshi')
glyph = font.createChar(-1, ligature_name)
glyph.addPosSub('ligatureshi', ligature_tuple)
glyph.build()
# below unlinking doesn't work, outputs nothing and stays linked
# , but it's not a problem for now. I don't probably need to unlink anyway
glyph.unlinkThisGlyph()
# Doing it glyph.unlinkRef('f_y') it says "OSError: No reference named f_y found in glyph f_y

font.save('ant.sfd')
#font.generate('ant1.otf', flags='opentype')
font.generate('ant1.otf', flags=('opentype',))
```

With above script I am able to create a new ligature f_y and build its components to it and see it in the metrics, then generate an otf font from it using the fontforge GUI. And yeah, only thing left now almost is that I want to be able to manipulate where it builds the f and y charaters. I want them to be so that `f` is on top and `y` is below it and that's it.

For the sake of what has happened here's some problems but they are maybe not important. 

1. Even if I specify the flag `opentype` in font.generate() it still doesn't generate that new table. Maybe the problem is that also it wasn't generating any ligatures in the table, and when saving to otf, it won't include empty tables and subtables?

2. Related to the above problem, I couldn't get the python script in command line to create ligatures in the table/subtable even in the saved .sfd file.

Solution to above: When I run the above script inside the fontforge GUI with the "Execute script", I am able to create ligatures in those table/subtable and that's fine for me so no need to solve those 1. and 2., just need to know how to move `f` on top and `y` below it using script

**Reply by khlsvr** [reply]
Continuing the "Trying to get the ligature component 1 to appear on top and ligature component 2 below it."  For example if I wanted to make a ligature of `+` and `-`, I want it to show up as `±`. 

The process creating a ligature ends in performing the `build -> build composite glyph` or running the python command `glyph.build()`. By default when I do that, it places them next to each other as if they were typed consequently. I can't find how I can manipulate the positions of these components programmatically so that they appear top-bottom.

I tried something by creating variants for a and b at new unicode slots called a.top and b.bottom. I created them so that a.top appears higher and b.bottom lower in the glyph area. For `a` glyph I added a vertical variant and put start/end/full values to something like 111/222/333. Did the same for `b` glyph. I then created a new glyph called a.top_b.bottom and set it as a ligature for a.top and b.bottom. All good thus far, but I can't build it. The `build` menu is grayed out. Also in the metrics if I type in /a.top_b.bottom I see nothing, but once I remove the last "m" I can see the a.top's glyph followed by the text "b.botto", and once I type "m" to complete the ligature, the metrics screen goes blank.

Before all that I also tried adjusting the positions desperately with `glyph.addPosSub(subtable_name, xoff, yoff, xadv, yadv)` in the script like `glyph.addPosSub(ligaterushi, 3, 5, 6, 8)` but it said that the `type expects 2 parameters, 5 given` or something like that. 

I have tried something else too but I'm lost at this point.

Edit: I have one idea after looking at docs, but I'm not sure ~~if this works or~~ if it's the way to go. I open the ligature glyph for a_b and then paste contents of a.top and b.bottom to it instead of running that glyph.build()

```
pen = font["a_b"].glyphPen(replace=False);		# Create a pen to draw into glyph "a_b"
font["a.top"].draw(pen);
font["b.bottom"].draw(pen);
pen = None;
```
